api/src/database/migrations/20210903A-add-auth-provider.ts

```diff
@@ -1,14 +1,21 @@
 import { Knex } from 'knex';
+import { getHelpers } from '../helpers';
 
 export async function up(knex: Knex): Promise<void> {
+	const helper = getHelpers(knex).schema;
+
 	await knex.schema.alterTable('directus_users', (table) => {
 		table.dropUnique(['email']);
 	});
 
 	await knex.schema.alterTable('directus_users', (table) => {
 		table.string('provider', 128).notNullable().defaultTo('default');
 		table.string('external_identifier').unique();
-		table.string('email', 128).nullable().alter();
+	});
+
+	await helper.changeToString('directus_users', 'email', {
```

Comment:
For this I think you need the "setNullable" and "dropNullable", right?

Reply:
I believe it does use that. `changeToString` calls `changeToTypeByCopy`:

https://github.com/directus/directus/blob/21565ed0bf4e8ce0a225565d415ddf9c2dcf522a/api/src/database/helpers/schema/dialects/oracle.ts#L17-L19

`nullable` option then calls 

https://github.com/directus/directus/blob/21565ed0bf4e8ce0a225565d415ddf9c2dcf522a/api/src/database/helpers/schema/types.ts#L118-L120

which uses setNullable/dropNullable

https://github.com/directus/directus/blob/21565ed0bf4e8ce0a225565d415ddf9c2dcf522a/api/src/database/helpers/schema/types.ts#L13-L19

Reply:
I was more assuming the whole copy process might be causing the error when it would be simpler to just do a "setNullable". Though really I'm just throwing stuff against the wall here.

Reply:
Ahh, because the default value of the temp column might already be nullable: true, at which point we try to make the renamed temp nullable again 🤔

Reply:
Actually, that looks like a bug in the `changeToByCopy` function? The copy builds the "temp" column with the options field, so it could be `nullable: true`. It'll then try and set the nullability again lower down?

https://github.com/directus/directus/pull/10113/files/21565ed0bf4e8ce0a225565d415ddf9c2dcf522a#diff-9f6662d181bc16992cd202060cb6f308887a2231a127516fd7704cc84dde951cR101

I think there is also a chance for errors here if the temp column is created with `nullable: false` in a table with data, since it'll fail with a "column cannot be null" error?

Reply:
Ignore my deleted incoherent rambling. I need to go to bed. 😅

Reply:
Rambling aside, the bug is that is that is you send `nullable: true` as an option and your `cb` leaves the column as default nullable, then it'll still try and set the column to nullable again here:

https://github.com/directus/directus/blob/21565ed0bf4e8ce0a225565d415ddf9c2dcf522a/api/src/database/helpers/schema/types.ts#L118

Should be something like `if (options.nullable === false) { builder.dropNullable(column); }`

Reply:
That being said, that would also fail if your `cb` decided to make the column not nullable. Someone less tired that me can figure it out.